### ISSUE TITLE
Support returning secrets from functions

### ISSUE THREAD

**Issue by t0yv0** [pulumi/pulumi#12710]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Providers may define function results as secret in the schema, in particular Terraform bridged providers can describe function outputs as sensitive, but Pulumi language SDKs currently are unable to work with these. The only workaround for providers is to elide secrets from the Invoke response, which can compromise some safety expectations, and if the user is not careful, and store the results of the function calls in the statefile, they may end up being saved in plaintext.

#### Golang

In the case of Go, receiving secret results from an Invoke leave the user with an unhelpful error message:

"unexpected secret result returned to invoke call"

https://github.com/pulumi/pulumi/blob/master/sdk/go/pulumi/context.go#L363

What would be helpful instead is to direct the user to use the "output-versioned invoke" version that is able to parse and represent the secrets correctly. 

For example in AWS there is LookupUserOuput form that should be able to represent the secrets:

https://github.com/pulumi/pulumi-aws/blob/master/sdk/go/aws/iam/getUser.go#L77

Unfortunately this requires some SDK changes as the underlying implementation code still uses the same secret-unfriendly cod path.

#### YAML 

In the case of YAML, the user gets an unhelpful message:

```
error: unexpected secret result returned to invoke call
```

YAML should be able to handle the secret value.

#### C# 
TBD

#### Python
TBD

#### Node
TBD

#### JAVA
TBD

**Comment by justinvp** [comment]
If a function returns a sensitive value, would we make the non-Output variant error, pointing to the Output variant? Or emit a warning but still allow it to succeed.

Note that it appears in some SDKs we do not generate an Output variant if the function does not have any inputs. This is tracked by https://github.com/pulumi/pulumi/issues/12449

**Comment by t0yv0** [comment]
> if a function returns a sensitive value, would we make the non-Output variant error, pointing to the Output variant? Or emit a warning, but still allow the function to return successfully

Great question. Tradeoffs.. I'm not sure which is right for us.

Something toward using a warning (as opposed to hard fail): programs that depend on plain invokes that return sensitive values may not be storing these values in state at the moment, and are not really *affected* in that sense, but erroring out would break those programs.

**Comment by Frassle** [comment]
> Great question. Tradeoffs.. I'm not sure which is right for us.

Sounds like we're going to be turning on "secretness" for invokes that are currently in use so I suspect we need to go with just warning.

**Comment by t0yv0** [comment]
@pierskarsenbarg mentioned https://www.pulumi.com/registry/packages/gcp/api-docs/secretmanager/getsecretversion/ as another use case here

**Comment by Sarabadu** [comment]
@t0yv0 this i think its quite simmilar to the `listStorageAccountKeys` 

this should also cover [`listWebAppHostKeys`](https://www.pulumi.com/registry/packages/azure-native/api-docs/web/listwebapphostkeys/) right? we are now wrapping the returned values with a `pulumi.secret` but would be nice to get the valueas as secrets by default

is this the right place to report it? or should we create a separate issue for it?

**Comment by t0yv0** [comment]
Yea this issue is good to upvote and x-ref with more use cases.

**Comment by Sarabadu** [comment]
adding some context to help prioritise too:
We ended up here because using the pulumi gh action with `preview` and `comment-on-pr` ended up disclosing multiple secrets in the gh comment. using `listStorageAccountKeys` and `listWebAppHostKeys`.

**Comment by t0yv0** [comment]
Checking up on this, unfortunately current implementation as of Aug 2024 still is not good enough to support secrets even when the user is using the Output-versioned invoke forms:

https://github.com/pulumi/pulumi-aws/blob/master/sdk/nodejs/secretsmanager/getSecretVersion.ts#L127

They're currently a thin veneer over the plain forms and that needs revisiting before we can enable providers to return secret material in the call.

**Comment by iwahbe** [comment]
We will need to carefully consider the desired behavior of calling a plain invoke where a secret value is returned by the provider.

On option (as brought up in https://github.com/pulumi/pulumi-go-provider/issues/260) is to allow the schema author to disable plain invokes where the provider knows a secret may be returned.

**Comment by t0yv0** [comment]
That is the way yes. These will need to be removed through deprecation + removing in next major. We might need to account for some plumbing to help provider authors manage this deprecation + transition.

**Comment by iwahbe** [comment]
@Zaid-Ajaj Do output-secrets in functions now work for all langauges? The PR that closed this only mentioned Python and Node.

**Comment by Zaid-Ajaj** [comment]
@iwahbe Python and Node were implemented after Go and YAML, here are the relevant PRs
 - For Go https://github.com/pulumi/pulumi/pull/17132
 - For YAML https://github.com/pulumi/pulumi-yaml/pull/622

Investigated C# and Java SDKs and determined that they already were keeping track of the secrets

**Comment by t0yv0** [comment]
I think this issue was incorrectly closed and there is stil
